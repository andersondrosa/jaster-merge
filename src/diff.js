function isObject(o) {
  return o instanceof Object && o.constructor.name == "Object";
}

const NO_DIFF = "&nodiff;";

function _diff(scope, original, changed) {
  //
  scope.i++;

  if (!isObject(original) || !isObject(changed)) {
    return changed;
  }

  const diff = {};

  for (const key in changed) {
    //
    if (!original.hasOwnProperty(key)) {
      diff[key] = changed[key];
      continue;
    }

    if (original[key] === changed[key]) {
      continue;
    }

    if (isObject(original[key]) && isObject(changed[key])) {
      const c = _diff(scope, original[key], changed[key]);

      if (c !== NO_DIFF) {
        diff[key] = c;
      }

      continue;
    }

    if (Array.isArray(original[key]) && Array.isArray(changed[key])) {
      if (JSON.stringify(original[key]) === JSON.stringify(changed[key])) {
        continue;
      }
    }

    diff[key] = changed[key];
  }

  for (const key in original) {
    if (!changed.hasOwnProperty(key)) {
      if (key.substr(0, 2) === "--") {
        continue;
      }

      if (!diff.hasOwnProperty("__rm")) {
        diff["__rm"] = [];
      }

      diff["__rm"].push(key);

      continue;
    }
  }

  for (const k in diff) {
    return diff;
  }

  return NO_DIFF;
}

export default function diff(a, b) {
  //
  const scope = { i: 0 };

  const diff = _diff(
    scope,
    JSON.parse(JSON.stringify(a)),
    JSON.parse(JSON.stringify(b))
  );

  if (diff === NO_DIFF) {
    return {};
  }

  return diff;
}
