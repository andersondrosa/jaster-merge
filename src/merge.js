function isObject(o) {
  return o instanceof Object && o.constructor.name == "Object";
}

function emptyObject(o) {
  for (let i in o) {
    return false;
  }
  return true;
}

function deepMerge(data, payload, scope) {
  //
  const scope_path = scope.path ? scope.path + "." : "";

  if (!isObject(data) || !isObject(payload)) {
    return false;
  }

  if (payload.hasOwnProperty("__rm")) {
    const rm = payload["__rm"];

    for (var i in rm) {
      const key = rm[i];

      if (data.hasOwnProperty(key)) {
        scope.data[key] = data[key];

        scope.env.deleted_keys.push(scope_path + key);

        delete data[key];
      }
    }

    delete payload["__rm"];
  }

  // -------------------------------------------------------------------------

  const remove = [];

  for (const key in payload) {
    //
    const row = payload[key];
    
    let _new = false;

    if (!data.hasOwnProperty(key)) {
      remove.push(key);

      scope.env.created_keys.push(scope_path + key);

      data[key] = row;

      continue;
    }

    // -----------------------------------------------------------------------

    if (isObject(row)) {
      if (row.hasOwnProperty("<:new:>")) {
        _new = true;
        delete row["<:new:>"];

        if (row.hasOwnProperty("__rm")) {
          delete row["__rm"];
        }
      }

      if (!_new && isObject(data[key])) {
        scope.data[key] = {};

        deepMerge(data[key], row, {
          env: scope.env,
          path: scope_path + key,
          data: scope.data[key]
        });

        continue;
      }
    }

    // -----------------------------------------------------------------------

    if (data[key] === row) {
      continue;
    }

    scope.data[key] = data[key];

    if (_new) {
      scope.data[key]["<:new:>"] = true;
    }

    scope.env.changed_keys.push(scope_path + key);

    data[key] = row;
  }

  if (remove.length) {
    scope.data["__rm"] = remove;
  }

  return true;
}

export default function merge(data, payload) {
  //
  const timestamp = new Date().getTime();

  // CÃ³pia de Payload
  payload = JSON.parse(JSON.stringify(payload));

  const env = {
    created_keys: [],
    changed_keys: [],
    deleted_keys: []
  };

  if (emptyObject(payload)) {
    return env;
  }

  const scope = {
    env,
    path: "",
    data: {}
  };

  deepMerge(data, payload, scope);

  const response = {
    rollback: scope.data,
    timestamp
  };

  let changes;

  if (env.created_keys.length) {
    changes = true;
    response["created_keys"] = env.created_keys;
  }

  if (env.changed_keys.length) {
    changes = true;
    response["changed_keys"] = env.changed_keys;
  }

  if (env.deleted_keys.length) {
    changes = true;
    response["deleted_keys"] = env.deleted_keys;
  }

  if (!changes) {
    response["unchanged"] = true;
  }

  return response;
}
