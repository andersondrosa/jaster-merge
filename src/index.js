import diff from "./diff";
import merge from "./merge";

export { diff, merge };
