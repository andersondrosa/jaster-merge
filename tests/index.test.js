import { diff, merge } from "../src";
it("Should be ok", function() {
  //
  // expect(diff).toMatchSnapshot();
  let data = {
    foo: "foo",
    bar: "bar"
  };

  let $merge = { foo: "FOO" };

  let { rollback } = merge(data, $merge);

  expect(data).toMatchSnapshot();
  
  merge(data, rollback)
  
  expect(data).toMatchSnapshot();

  const $diff = diff(data, $merge);

  expect($diff).toMatchSnapshot();
});
