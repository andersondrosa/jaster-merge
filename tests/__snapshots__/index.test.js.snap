// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should be ok 1`] = `
Object {
  "bar": "bar",
  "foo": "FOO",
}
`;

exports[`Should be ok 2`] = `
Object {
  "bar": "bar",
  "foo": "foo",
}
`;

exports[`Should be ok 3`] = `
Object {
  "__rm": Array [
    "bar",
  ],
  "foo": "FOO",
}
`;
